src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -0,0 +1,1555 @@
+
+  ceph_assert(ceph_mutex_is_locked_by_me(m_lock));
+
+  auto itr = std::find_if(
+      m_remote_group_snaps.begin(), m_remote_group_snaps.end(),
+      [group_snap_id](const cls::rbd::GroupSnapshot &s) {
+      return s.id == group_snap_id;
+      });
+
+  ceph_assert(itr != m_remote_group_snaps.end());
+  auto itl = std::find_if(
+      m_local_group_snaps.begin(), m_local_group_snaps.end(),
+      [group_snap_id](const cls::rbd::GroupSnapshot &s) {
+      return s.id == group_snap_id;
+      });
+  if (itr->snaps.size() != 0) {
+    // update the group snap with snap spec
+    itl->snaps.push_back(*spec);
+  }
+
+  if (itr->snaps.size() == itl->snaps.size()) {
+    itl->state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+
+    auto comp = create_rados_callback(
+        new LambdaContext([this, group_snap_id, on_finish](int r) {
+          handle_mirror_snapshot_complete(r, group_snap_id, on_finish);
+          }));
+
+    librados::ObjectWriteOperation op;
+    librbd::cls_client::group_snap_set(&op, *itl);
+    int r = m_local_io_ctx.aio_operate(
+        librbd::util::group_header_name(m_local_group_id), comp, &op);
+    ceph_assert(r == 0);
+    comp->release();
+  } else {
+    on_finish->complete(0);
+  }
+
+  dout(10) << "local group snap info: "
+           << "id: " << itl->id
+           << ", name: " << itl->name
+           << ", state: " << itl->state
+           << ", snaps.size: " << itl->snaps.size()
+           << dendl;
+
+  return;
+}
+
+template <typename I>
+void Replayer<I>::handle_mirror_snapshot_complete(
+    int r, const std::string &group_snap_id, Context *on_finish) {
+  dout(10) << group_snap_id << ", r=" << r << dendl;
+
+  if (r < 0) {
+    on_finish->complete(r);
+    return;
+  }
+
+  utime_t duration = ceph_clock_now() - m_snapshot_start;
+  m_last_snapshot_complete_seconds = duration.sec();
+  m_snapshot_start = utime_t(0, 0);
+
+  uint64_t last_snapshot_bytes = 0;
+  for (const auto& ir : *m_image_replayers) {
+    if (ir.second != nullptr) {
+      last_snapshot_bytes += ir.second->get_last_snapshot_bytes();
+    }
+  }
+  m_last_snapshot_bytes = last_snapshot_bytes;
+
+  on_finish->complete(0);
+}
+
+template <typename I>
+void Replayer<I>::remove_mirror_peer_uuid(const std::string &snap_id) {
+  auto remote_snap = std::find_if(
+      m_remote_group_snaps.begin(), m_remote_group_snaps.end(),
+      [snap_id](const cls::rbd::GroupSnapshot &s) {
+      return s.id == snap_id;
+      });
+
+  if (remote_snap == m_remote_group_snaps.end()) {
+    derr << "remote group snapshot not found: "
+         << snap_id << dendl;
+    return;
+  }
+
+  auto rns = std::get_if<cls::rbd::GroupSnapshotNamespaceMirror>(
+      &remote_snap->snapshot_namespace);
+  if (rns == nullptr) {
+    derr << "remote group snapshot is not a mirror snapshot: "
+         << snap_id << dendl;
+    return;
+  }
+
+  if (!rns->mirror_peer_uuids.empty() &&
+      rns->mirror_peer_uuids.count(m_remote_mirror_peer_uuid) != 0) {
```

Comment:
The emptiness check seems redundant:
```suggestion
  if (rns->mirror_peer_uuids.count(m_remote_mirror_peer_uuid) != 0) {
```

Reply:
Done.